configuration/pih/liquibase/sql/create_functions.sql

```diff
@@ -835,6 +857,83 @@ BEGIN
 
     RETURN enc_id_out;
 
+END
+#
+/*
+This function accepts a patient id, source & code of a concept and a locale
+It will return the single, latest value coded of the obs of that concept in the locale
+*/
+DROP FUNCTION IF EXISTS latest_obs_value_coded;
+#
+CREATE FUNCTION latest_obs_value_coded(_patient_id int(11), _source varchar(50), _term varchar(255), _locale varchar(50))
+    RETURNS text
+    DETERMINISTIC
+
+BEGIN
+
+    DECLARE ret text;
+
+    select      concept_name(o.value_coded, _locale) into ret
+    from        obs o
+    where       o.voided = 0
+		and o.person_id = _patient_id    		
+      	and       o.concept_id = concept_from_mapping(_source, _term)
+    order by obs_datetime desc limit 1 ;
+
+    RETURN ret;
+
+END
+#
+/*
+This function accepts a patient id, source & code of a concept
+It will return the single, latest numeric value of the obs of that concept in the locale
+*/
+#
+DROP FUNCTION IF EXISTS latest_obs_value_numeric;
+#
+CREATE FUNCTION latest_obs_value_numeric(_patient_id int(11), _source varchar(50), _term varchar(255))
+    RETURNS double
+    DETERMINISTIC
+
+BEGIN
+
+    DECLARE ret double;
```

Comment:
Similar to my comment above, if you are going to have lots of these functions, you might want to consider having:

latest_obs(_patient_id, _source, _term) -> returns obs_id

and then:

value_numeric(_obs_id)
value_datetime(_obs_id)
value_coded(_obs_id);
value_coded_name(_obs_id);
obs_datetime(_obs_id);
etc.

This would also allow you have additional functions like these, and compose the results with the same set of obs functions:

first_obs(_patient_id, _source, _term);
latest_obs(_patient_id, _source, _term, _index_desc);
earliest_obs(_patient_id, _source, _term, _index_asc);

Reply:
I like this approach. will rework as described.